nixos/modules/services/web-apps/lotide.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.lotide;
+in
+{
+  options.services.lotide = {
+    enable = mkEnableOption "the lotide service";
+
+    package = mkOption {
+      default = pkgs.lotide;
+      defaultText = "pkgs.lotide";
+      type = types.package;
+      description = "lotide package to use.";
+    };
+
+    databaseUrl = mkOption {
+      default = null;
+      type = types.string;
+      description = "Credentials for database connection.";
+      example = "postgresql://user:passwort@somehost:5432/databasename";
+    };
+
+    hostUrlActivityPub = mkOption {
+      default = null;
+      type = types.string;
+      description = "If using the recommended proxy setup, set this to your root address";
+      example = "https://example.com";
+    };
+
+    hostUrlAPI = mkOption {
+      default = null;
+      type = types.string;
+      description = "URL for API";
+
+      example = "https://example.com/api";
+    };
+
+    apubProxyRewrites= mkOption {
+      default = false;
+      type = types.bool;
+      description = "Set to `true` to make signatures work with the proxy setup.";
+    };
+
+    allowForwarded = mkOption {
+      default = null;
+      type = types.string;
+      description = "Set to true to make ratelimiting work with the proxy setup.";
+    };
+
+    smtpUrl = mkOption {
+      default = null;
+      type = types.string;
+      description = "URL used to access SMTP server, required for sending email";
+
+      example = "smtps://username:password@smtp.example.com";
+    };
+
+    smtpFrom = mkOption {
+      default = null;
+      type = types.string;
+      description = "From value used in sent emails, required for sending email";
+    };
+
+    mediaLocation = mkOption {
+      default = null;
+      type = types.string;
+      description = ''
+        Directory on disk used for storing uploaded images. If not set, image uploads will be disabled.
+        Must be created before this service is started.
+      '';
+
+      example = "/var/lotide/imgs";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.lotide = {
```

Comment:
Would you consider submitting a change upstream to include a `systemd` unit in their repo? If so, we could eventually (mostly) drop the unit from nixpkgs an utilize `systemd.packages = [ pkgs.lotide ];`. If you're up for that full bonus points will be awarded :tada:

Reply:
Can you give some more detail on that? Does that mean that the `lotide.service` or `hitide.service` file from upstream can be included here? How can I point to the actual path in the upstream repository? How can I make them configurable from nix?

Reply:
Sure. If `pkgs.lotide` any `.service` files under `$out/lib/systemd/system` and you specify `systemd.packages = [ pkgs.lotide ];` then these `systemd` unit files will be included in your system when built.

Checkout [this](https://github.com/NixOS/nixpkgs/pull/100063/commits/4f5d3794d3348fd7da1c3954ef9ef80d4d382acc) commit where I made use of the upstream `systemd` unit, but then provided a few changes on top of the upstream unit specifically for NixOS.

Reply:
So the package needs to install them properly, okay.
And there's no way for a user of the service definition to overwrite values in the definition?

Reply:
I don't think so.

Reply:
Hm. After some more thinking about the whole thing, I guess it is not a good idea to include a .service file from upstream - because lotide/hitide is configured via environment variables... A user actually _wants_ to overwrite the environment of the services. Having a .service file prevents that (from my understanding).

Reply:
Both NixOS and `systemd` excel here. Both allow extensive overriding to handle situations like this, both from the perspective of the distro and the perspective of the sysadmin.

Here is a reasonable starting point to learn about this: https://serverfault.com/questions/413397/how-to-set-environment-variable-in-systemd-service

Let me know if you have any questions.